pkg/daemon/ceph/osd/kms/envs.go

```diff
@@ -63,30 +79,48 @@ func vaultTLSEnvVarFromSecret(kmsConfig map[string]string) []v1.EnvVar {
 	return vaultTLSEnvVar
 }
 
-// VaultConfigToEnvVar populates the kms config as env variables
-func VaultConfigToEnvVar(spec cephv1.ClusterSpec) []v1.EnvVar {
+// ConfigToEnvVar populates the kms config as env variables
+func ConfigToEnvVar(spec cephv1.ClusterSpec) []v1.EnvVar {
 	envs := []v1.EnvVar{}
-	backendPath := GetParam(spec.Security.KeyManagementService.ConnectionDetails, vault.VaultBackendPathKey)
-	// Set BACKEND_PATH to the API's default if not passed
-	if backendPath == "" {
-		spec.Security.KeyManagementService.ConnectionDetails[vault.VaultBackendPathKey] = vault.DefaultBackendPath
+
+	if spec.Security.KeyManagementService.IsVaultKMS() {
+		backendPath := GetParam(spec.Security.KeyManagementService.ConnectionDetails, vault.VaultBackendPathKey)
+		// Set BACKEND_PATH to the API's default if not passed
+		if backendPath == "" {
+			spec.Security.KeyManagementService.ConnectionDetails[vault.VaultBackendPathKey] = vault.DefaultBackendPath
+		}
 	}
+
+	if spec.Security.KeyManagementService.IsIBMKeyProtectKMS() {
+		// We don't want to leak the IBM service API key to the container environment variables even
+		// the container is ephemeral.
+		// The IBM_SERVICE_API_KEY content is mounted in the provisioner container as an
+		// environment variable from a secret
+		delete(spec.Security.KeyManagementService.ConnectionDetails, ibm.IbmServiceApiKey)
+		envs = append(envs, ibmKeyProtectServiceAPIKeyEnvVarFromSecret(spec.Security.KeyManagementService.TokenSecretName))
+	}
+
 	for k, v := range spec.Security.KeyManagementService.ConnectionDetails {
-		// Skip TLS and token env var to avoid env being set multiple times
-		toSkip := append(cephv1.VaultTLSConnectionDetails, api.EnvVaultToken)
-		if sets.NewString(toSkip...).Has(k) {
-			continue
+		if spec.Security.KeyManagementService.IsVaultKMS() {
```

Comment:
Because the Vault implementation uses the libopenstorage lib which understand env variables and the kp implementation uses the native go-client from IBM.